### ISSUE TITLE
Question: Dynamic Services on Startup

### ISSUE THREAD

**Issue by alexyao2015** [just-containers/s6-overlay#408]
Is creating x number of services on startup using an init script possible? Specifically, I want to pass x number of environment variables, use an init script to parse the variables and create x number of services based on the environment variables, and then monitor the services generated.

To provide context, I'm trying to run multiple instances of ffmpeg as a service with each of the instance arguments provided in the environment variable.

**Comment by skarnet** [comment]
Instances are a long-standing feature wish in s6, but they're deceptively hard to implement correctly, especially dynamic instances which is the thing you're asking for. Ultimately they will be implemented in the s6-rc service manager, but it will yet take some time. So, unfortunately, as of yet there isn't any native support for instances.

However, in s6-overlay, if you are running your container with a read-write rootfs, you can do this manually: in a `/etc/cont-init.d` script, clear `/etc/services.d`, then dynamically create in `/etc/services.d` all the services you need based on your environment variables.
Basically, it means treating `/etc/services.d` as a dynamically generated longshot repository, instead of a static one. It's not ideal, but it will work.

**Comment by alexyao2015** [comment]
Thanks for the quick response and help! 

Is this possible to do with the new service directory structure in v3 instead of the legacy implementation? 

On a slightly different topic, I see the new service implementation has support for dependencies. Does this support a service dependency as a one shot service as well? Effectively I have:
1. A service that checks if the db is up
2. A one shot services that configured the db when it's up
3. The actual app that depends on 2

Lastly, I had used something [like this](https://github.com/hass-emulated-hue/core/blob/master/docker/rootfs/etc/cont-init.d/00-set-vars.sh#L12) in the past to modify the container environment using an init script. I know this is slightly hacky, but I would like to understand if there is a better way to achieve this or if it should simply be avoided entirely.

Thanks again for all your hard work and for reviving s6-overlay!

**Comment by alexyao2015** [comment]
Awesome thank you for the help! 

I had actually been using s6-notifyoncheck in v2, so this fits along the same line in v3.

In the case of using the legacy service directory, is [this](https://github.com/zoneminder-containers/zoneminder-base/blob/main/rootfs/etc/services.d/mariadb/run#L11) still the correct syntax to wait for a dependency to startup?

**Comment by skarnet** [comment]
Yes, it is. However, I don't think your use of s6-notifyoncheck on the previous line is correct. You should only be using s6-notifyoncheck to signal readiness of the *current* service, not of another one. If you want to call the checking program, invoke `/var/run/s6/services/mariadbdata/check` directly.

Also, after that, you poll every second, running the checking program until it fails. The point of supervision is that you don't need to do that: the `finish` script for your mariadb service is the place to do whatever you need to do when mariadb dies.

All in all, you really don't need that stub service: you can move your mariadb service into s6-rc with readiness notification and dependencies on it directly, and a finish script to handle its failure, and save yourself a lot of complexity and polling.

**Comment by alexyao2015** [comment]
I'm not sure I fully understand why this would be incorrect. 

To provide some context on that project(not yet updated for v3 yet), this polls an external db (not running inside the container), so there is no long-running process within that "service". The polling every second is there to simulate a long-running service and terminates when the external service dies to allow the `finish` script to terminate the supervision tree. The initial readiness notification allowed the other depending services to start.

I have not tested this yet, but in v2, there existed a bug where terminating the supervision tree did not send a sigterm to processes not supervised under a service directly. This caused the container stop to stall. Does this issue exist in v3? (I have a misbehaving app that uses a daemon to launch a bunch of other processes with no real way to bypass it.)

**Comment by skarnet** [comment]
Ah, if your mariadb service isn't running inside the container then the polling makes more sense. It would still be better, though, if you could have a client with an open connection to mariadb, blocking and doing nothing, until the mariadb server exits and force closes the connection, at which point your client dies. No polling would be necessary then. And you could notify readiness when the client first manages to connect to the server.

v3 hopefully fixes all the shutdown bugs that existed in v2, that was a major factor in the decision to rewrite s6-overlay. In v3, on container shutdown, the services are stopped in an orderly fashion, and after cont-finish scripts are run, every remaining process gets a SIGTERM, then a SIGKILL if necessary, before the container exits. So your misbehaving app should be properly handled now.

**Comment by alexyao2015** [comment]
Thats a great idea! I'll look into figuring that out. Thanks for your help with all my questions. Excited to get v3 implemented in my containers.

**Comment by alexyao2015** [comment]
@skarnet Quick question, I see the documentation for `s6-rc-compile` states:

```
Note that you cannot create a ./down file for [s6-supervise](https://skarnet.org/software/s6/s6-supervise.html) in a generated service directory.
```

In the past, I had used `s6-svc -o` to run down services as scheduled scripts using cron. It appears this is not possible with the new style service directory. Is that correct?

Additionally, I previously used a command such as `/bin/s6-svc -o /var/run/s6/services/script-myscript` to bootup down services using cron. It appears that in the v3, the temporary directory has been moved to `/run/s6/legacy-services`, and I can continue using my method by changing the above command. I would presume that the method I was previously using to run scheduled services was a little hacky. Is there any more official/supported way to run services on a schedule?

**Comment by alexyao2015** [comment]
One more question, is there a way to both use `notification-fd` and have the text also get written to stdout and `log`? If I use `fdmove -c 1 3`, STDOUT is redirected to fd 3 and no longer logs to the container STDOUT or the logger util. Is there a way to copy the inputs from one fd to another?

**Comment by skarnet** [comment]
You can continue using your method (and `/var/run` will even still work because s6-overlay makes sure it's a symlink to `/run`, for compatibility). However, you're right that it's a little hacky. If you're running periodic scripts, these scripts are actually not supervised services! `cron` is managing them, not s6; s6 doesn't really give you any significant benefit here. You could store the run script anywhere else, name it anything else, and have `cron` call it directly, instead of designing a service directory and having `cron` activate the service.
Alternatively, you could look into something like [snooze](https://github.com/leahneukirchen/snooze) which is a cron-like designed to run in integration with a supervision system such as s6.

The point of `notification-fd` is that it can be any fd, in particular it should not prevent you from using your stdout. Using something like `fdmove 1 3` is only good if you're using something like [`s6-tcpserver4 -1`](https://skarnet.org/software/s6-networking/s6-tcpserver4.html) that can only write its notification message to its stdout, and does not have a regular output stream otherwise.
Do not send a data stream to `notification-fd`, that isn't what it was made for and might make your service fail (because the writes will fail at some point). Only use `notification-fd` within its definition domain: when the service is ready, it should write something that ends with a newline to `notification-fd`, then never write anything to that fd again. (And it can close the fd.) Do you have anything in your service that could be used that way?

**Comment by alexyao2015** [comment]
> You can continue using your method (and /var/run will even still work

In the past, the services were located in `/var/run/s6/services`, while now they are located in `/var/run/service`, even with the symlink. 

> s6 doesn't really give you any significant benefit here

The main purpose of using services to do this is so that I can still have the logging features of using s6-log and sending the logs to the container STDOUT as well. I'm not sure if there is a more intuitive or simpler way to have these features otherwise. 

> because the writes will fail at some point

Ah, this explains why I would see some services fail and continually restart. This particular isn't documented in the s6 service directory page, and that would help a ton.

> Do you have anything in your service that could be used that way?

In particular, I was using `crond`, which like many services not built for s6, doesn't have a readiness notifications built-in except for logging to STDOUT.  Maybe this is a broader question, but how could I use `notification-fd` for a program that doesn't have readiness notifications built-in such as `crond`? 

As you mentioned earlier, `you could notify readiness when the client first manages to connect to the server`, it seems in order to achieve this, I would need to create my own client/wrapper/standalone script to notify on readiness rather than executing the client directly through fdmove.

**Comment by alexyao2015** [comment]
Thank you again for the very clear explaination.

Sorry for turning this into a Q&A thread. If there is a better place to have this conversation, please let me know.

I've gone ahead and created my first project using the new service directories, but I seem to be having an issue with the logging pipelines. In the past, I was able to using `log/run` to append text to the log, log to STDOUT, and log to s6-log. Here, it seems that only my `longrun` services are being logged using the pipeline to s6-log, however, the logs are not being sent to STDOUT as well. Am I supposed to have the formatter exist as one pipeline with two other services to 1. dump to STDOUT and 2. log with s6-log? It seems that this would tremendously clutter service definitions here.

Services marked as `oneshot` (where the logger is forced to also be oneshot as the producer and consumer have to be the same type) appear the same as not having a logger at all where logs are directly dumped to STDOUT with no formatting. The log folders and files created by s6-log exist but are entirely empty.

Is there a simple way to achieve logging [as I have in the past](https://github.com/alexyao2015/cloud-media-scripts/blob/refactor/root/etc/services.d/cloud-decrypt/log/run)? 

[Here is the source of the project I am referring to.](https://github.com/alexyao2015/nfs-server-docker/tree/master/root/etc/s6-overlay/s6-rc.d) Hopefully, this helps to demonstrate visually my observations.

**Comment by skarnet** [comment]
For support about s6 or s6-rc that is not specific to s6-overlay, it's probably better to ask on one of the [skarnet.org mailing-lists](https://skarnet.org/lists/); or, if you can, come to the `#s6` channel on the OFTC IRC network and ask your questions there; the community will likely answer you even if I'm not here.

As to your questions:

- Indeed, only longruns use pipelines under s6-rc. The point of pipelines is to maintain the pipes open even when one process in the pipeline dies (this avoids losing data and having to kill the whole chain of processes); for oneshots, which are by definition short-lived, this is unnecessary. `producer-for` and `consumer-for` files are meaningless for oneshots, so it's expected that you didn't see anything in your logs. :-)
- You can still pipe the output of your oneshot up script to a logger, the logger invocation just needs to be a part of your script.
  * If you want to do this without modifying your oneshot script, have your `up` script be the following: `pipeline -w { your logger command line } your_oneshot_script`.
- I am surprised that your previous version of logging did actually log to stdout as well as to your logging directory. The script you linked certainly does not indicate that it's doing this.
- s6-log *can* log to stdout, but you need to give it the `1` directive. It's not possible to do so via `logutil-service`, but you can write your `s6-log` command line by hand to add the instruction. (It's easy: look at the `ps` output for your current logger, copy the `s6-log` command line, and add a `1` at the end.) Then you should have the duplicated logs you need.
- Rather than having a `|` in your logger, it is more idiomatic to have a longer pipeline (s6-rc can do that) with only one operation per service. So you could declare three longruns:
  * Your service, a producer for the formatter
  * Your formatter (the `s6-format-filter` invocation), a consumer for the service and a producer for the logger
  * Your logger (the `s6-log` invocation with the `1` directive to also log to stdout), a consumer for the formatter.
- wrt fix-attrs: are you getting a message that says fix-attrs is being applied to your file, and warning you that fix-attrs is deprecated? If you're getting no messages at all then fix-attrs is not even reading your file. If you are getting messages, it means the problem happens in the fix-attrs script itself.

**Comment by alexyao2015** [comment]
> it's probably better to ask on one of the skarnet.org mailing-lists

I've attempted to join the mailing list, but I don't believe my join has been approved?

> It's not possible to do so via logutil-service

I figured out that it is possible by setting the `S6_LOGGING_SCRIPT` variable, and that my previous script worked by including 1 in this script. I had totally forgotten to set this!

> are you getting a message that says fix-attrs is being applied to your file, and warning you that fix-attrs is deprecated?

Yes I am. There appears to be every indication that fix-attrs is working, however, the directory is unmodified. There are no errors produced, and I do see the deprecation message. I would expect that [this fix-attrs](https://github.com/alexyao2015/cloud-media-scripts/blob/935f70c6cf9de311a2ee423801bee5cc1f04d0f1/root/etc/fix-attrs.d/00-fix-permissions) file is the equivalent to [this startup script](https://github.com/alexyao2015/cloud-media-scripts/blob/935f70c6cf9de311a2ee423801bee5cc1f04d0f1/root/etc/cont-init.d/00-fix-permissions.sh). Either way, this is not too big of a deal as this can still be achieved easily using a startup script. 

Lastly, I seem to be running into a benign warning/error that crops up in the logs `: No such file or directory: unable to exec sh` for my longrun loggers that seem to be coming from the shebang. This exact same error comes up both when using `#!/command/with-contenv sh` and `#!/bin/sh`.

**Comment by skarnet** [comment]
- Your join has been approved now.
- Yes, if you add `1` to `S6_LOGGING_SCRIPT` it will make s6-log log to stdout. But then it will apply to all your uses of `logutil-service`; I couldn't tell whether it's what you want. Also, `1` is an action directive, so make sure it's the *last* instruction in `S6_LOGGING_SCRIPT`, else the stdout output will miss the settings you declare after it.
- I will investigate the fix-attrs script, thanks for the report.
- Can you provide the whole run script of a logger that fails, and the contents of your `S6_LOGGING_SCRIPT`? This is a weird error, and may not be as harmless as it seems.

**Comment by alexyao2015** [comment]
See [here](https://github.com/alexyao2015/nfs-server-docker/blob/master/root/etc/s6-overlay/s6-rc.d/20-rpcbind-log/run) for the log script and [here](https://github.com/alexyao2015/nfs-server-docker/blob/master/Dockerfile#L54) for `S6_LOGGING_SCRIPT`. All my log scripts are the same. I'm actually this same script for my one shot loggers (through a pipeline in the up file) which do not generate this error.

**Comment by Mek101** [comment]
> Instances are a long-standing feature wish in s6, but they're deceptively hard to implement correctly, especially dynamic instances which is the thing you're asking for. Ultimately they will be implemented in the s6-rc service manager, but it will yet take some time. So, unfortunately, as of yet there isn't any native support for instances.

Has this been implemented since two years ago?

**Comment by skarnet** [comment]
Dynamic instances are now available in s6, yes: see the [s6-instance-maker](https://skarnet.org/software/s6/s6-instance-maker.html) page and the other commands it links to.

However, there's no syntactic sugar provided in s6-overlay for it, so calling `s6-instance-maker` should currently be manually done in a `S6_STAGE2_HOOK` script, and then instance management via `s6-instance-create` should be done in a later s6-rc service.

I may add syntactic sugar for easier instance management in the future but that's not a priority for now.

**Comment by raphaelcoeffic** [comment]
@skarnet Thx a lot for this new feature! It landed just I needed it 🤩

I'm installing my template services into `/etc/s6-overlay/templates` like this:
- `Dockerfile` excerpt:
```dockerfile
# Add services
COPY services /etc/s6-overlay/s6-rc.d
COPY service-templates /etc/s6-overlay/templates

ENV S6_STAGE2_HOOK=/usr/sbin/init-template-services.sh
```

- `/usr/sbin/init-template-services.sh`:
```bash
#!/bin/bash

# Service templates must be installed ealry in the game
# so s6-rc can pick them up and add them to the scanned
# directories.
#
# see https://github.com/just-containers/s6-overlay/issues/408#issuecomment-2239308186

TEMPLATE_DIR=/etc/s6-overlay/templates

make_instance() {
  /command/s6-instance-maker "$TEMPLATE_DIR/$1" "/run/service/$1"
}

for template in /etc/s6-overlay/templates/*; do
  template="$(basename $template)"
  make_instance "$template"
done
```

- later on, I then just to use this to create instances:
```
s6-instance-create /run/service/[template service name] [instance name]
```
- or this to delete them:
```
s6-instance-delete /run/service/[template service name] [instance name]
```

- in between, I'm using this to restart an instance:
```
s6-svc -r /run/service/[template service name]/[instance name]
```

*Please note: I did not bother trying out `s6-rc restart` to restart services, as my other scripts use `s6-svc` directly.*
